### PR TITLE
Add spiffe-csi driver support

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>
Co-authored-by: Furkan Türkal <furkan.turkal@trendyol.com>

/cc @dentrax

PTAL @marcofranssen 

You can test it with the following steps:

1. Deploy the example _client_ pod with the following manifest:

```yaml
cat << EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: client
  labels:
    app: client
spec:
  selector:
    matchLabels:
      app: client
  template:
    metadata:
      labels:
        app: client
    spec:
      hostPID: true
      hostNetwork: true
      dnsPolicy: ClusterFirstWithHostNet
      containers:
        - name: client
          image: gcr.io/spiffe-io/spire-agent:1.4.4
          command: ["sleep"]
          args: ["1000000000"]
          volumeMounts:
            - name: spiffe-workload-api
              mountPath: /spiffe-workload-api
              readOnly: true
      volumes:
        - name: spiffe-workload-api
          csi:
            driver: "csi.spiffe.io"
            readOnly: true
EOF
```

2. Now, you can shell into the _client_ pod and test it by fetching the SVID for it:

```shell
$ /opt/spire/bin/spire-agent api fetch -socketPath /spiffe-workload-api/agent.sock
Received 1 svid after 13.278918ms

SPIFFE ID:		spiffe://example.org/ns/default/sa/default
SVID Valid After:	2022-10-24 17:51:57 +0000 UTC
SVID Valid Until:	2022-10-24 18:52:07 +0000 UTC
CA #1 Valid After:	2022-10-24 17:50:42 +0000 UTC
CA #1 Valid Until:	2022-10-25 17:50:52 +0000 UTC
```

### CODE REVIEW

File: charts/spire/templates/agent-daemonset.yaml [hunk 1]
```diff
@@ -73,3 +122,23 @@ spec:
...
+        - name: mountpoint-dir
+          hostPath:
+            path: /var/lib/kubelet/pods
+            type: Directory
+        # This volume is where the node-driver-registrar registers the plugin
+        # with kubelet
+        - name: kubelet-plugin-registration-dir
+          hostPath:
+            path: /var/lib/kubelet/plugins_registry
+            type: Directory
```
Comment (marcofranssen) [inline charts/spire/templates/agent-daemonset.yaml:144]: Please end this file with a new line.

File: charts/spire/values.yaml [hunk 2]
```diff
@@ -9,14 +9,14 @@ workloadRegistrar:
     repository: gcr.io/spiffe-io/k8s-workload-registrar
     pullPolicy: IfNotPresent
     # Overrides the image tag whose default is the chart appVersion.
-    tag: ""
+    tag: "1.4.4"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:12]: I prefer to bump the appVersion in Chart.yaml and leave this empty.

File: charts/spire/values.yaml [hunk 3]
```diff
@@ -9,14 +9,14 @@ workloadRegistrar:
...
 
 server:
   image:
     repository: gcr.io/spiffe-io/spire-server
     pullPolicy: IfNotPresent
     # Overrides the image tag whose default is the chart appVersion.
-    tag: ""
+    tag: "1.4.4"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:19]: I prefer to bump the appVersion in Chart.yaml and leave this empty.

File: charts/spire/values.yaml [hunk 4]
```diff
@@ -34,7 +34,7 @@ oidc:
   image:
     repository: gcr.io/spiffe-io/oidc-discovery-provider
     pullPolicy: IfNotPresent
-    tag: ""
+    tag: "1.4.4"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:37]: I prefer to bump the appVersion in Chart.yaml and leave this empty.

File: charts/spire/values.yaml [hunk 5]
```diff
@@ -71,7 +71,7 @@ agent:
     repository: gcr.io/spiffe-io/spire-agent
     pullPolicy: IfNotPresent
     # Overrides the image tag whose default is the chart appVersion.
-    tag: ""
+    tag: "1.4.4"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:74]: I prefer to bump the appVersion in Chart.yaml and leave this empty.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Thanks a lot for this contribution. 🎉 

I'll give this a test as soon as possible. For now a few nit picky suggestions.

File: charts/spire-client-example/README.md [hunk 6]
```diff
@@ -1,7 +1,5 @@
 # spire-client-example
 
-<!-- This README.md is generated. -->
```
Comment (marcofranssen) [inline charts/spire-client-example/README.md:3]: Please us the template in the command.

https://github.com/philips-labs/helm-charts/blob/main/.github/README.md.tmpl

```shell
helm-docs -g charts/spire -t .github/README.md.tmpl
```

I should probably document this in a CONTRIBUTING.md

**Comment by developer-guy** [comment]
hello @marcofranssen, I just wanted to let you know that similar work has been done in the FRSCA project, I'm linking that issue.

> _https://github.com/buildsec/frsca/pull/355_

**Review by marcofranssen — APPROVED** [review]
LGTM

File: charts/spire/templates/agent-daemonset.yaml [hunk 7]
```diff
@@ -55,6 +55,55 @@ spec:
               port: 8080
             initialDelaySeconds: 15
             periodSeconds: 60
+        # This is the container which runs the SPIFFE CSI driver.
+        - name: spiffe-csi-driver
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply charts/spire/templates/agent-daemonset.yaml:59]: yeah, I'll do it in separate PR

File: charts/spire/templates/agent-daemonset.yaml [hunk 8]
```diff
@@ -55,6 +55,55 @@ spec:
...
+            - mountPath: /var/lib/kubelet/pods
+              mountPropagation: Bidirectional
+              name: mountpoint-dir
+          securityContext:
+            privileged: true
+        # This container runs the CSI Node Driver Registrar which takes care
+        # of all the little details required to register a CSI driver with
+        # the kubelet.
+        - name: node-driver-registrar
+          image: quay.io/k8scsi/csi-node-driver-registrar:v2.0.1
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply charts/spire/templates/agent-daemonset.yaml:93]: yeah, I'll do it in separate PR

**Pull request merged**